File: qqbot-client/src/main/java/com/mieai/qqbot/client/QqMediaDownloader.java
package com.mieai.qqbot.client;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Locale;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/** Bounded, redirect-aware downloader used before handing media to QQ. */
public final class QqMediaDownloader {
    private final HttpClient client;
    private final long maxBytes;
    private final Duration timeout;
    private final int maxRedirects;

    public QqMediaDownloader(long maxBytes, Duration timeout, int maxRedirects) {
        if (maxBytes < 1L) throw new IllegalArgumentException("maxBytes must be positive");
        this.maxBytes = maxBytes;
        this.timeout = Objects.requireNonNull(timeout, "timeout must not be null");
        if (timeout.isZero() || timeout.isNegative()) throw new IllegalArgumentException("timeout must be positive");
        if (maxRedirects < 0 || maxRedirects > 8) throw new IllegalArgumentException("maxRedirects is invalid");
        this.maxRedirects = maxRedirects;
        client = HttpClient.newBuilder().connectTimeout(timeout).followRedirects(HttpClient.Redirect.NEVER).build();
    }

    public CompletionStage<byte[]> download(URI source) {
        return CompletableFuture.supplyAsync(() -> downloadSync(source));
    }

    private byte[] downloadSync(URI source) {
        URI current = validate(source);
        for (int redirect = 0; ; redirect++) {
            HttpRequest request = HttpRequest.newBuilder(current).timeout(timeout)
                    .header("Accept", "*/*").GET().build();
            try {
                HttpResponse<InputStream> response = client.send(request, HttpResponse.BodyHandlers.ofInputStream());
                int status = response.statusCode();
                if (status >= 300 && status < 400) {
                    response.body().close();
                    if (redirect >= maxRedirects) throw new IllegalArgumentException("media URL exceeded redirect limit");
                    String location = response.headers().firstValue("location").orElseThrow(
                            () -> new IllegalArgumentException("media redirect has no Location header"));
                    current = validate(current.resolve(location));
                    continue;
                }
                if (status < 200 || status >= 300) {
                    response.body().close();
                    throw new IllegalArgumentException("media URL returned HTTP " + status);
                }
                long declared = response.headers().firstValueAsLong("content-length").orElse(-1L);
                if (declared > maxBytes) throw new IllegalArgumentException("media download exceeds configured limit");
                try (InputStream input = response.body(); ByteArrayOutputStream output = new ByteArrayOutputStream()) {
                    byte[] buffer = new byte[8192];
                    long total = 0L;
                    int read;
                    while ((read = input.read(buffer)) != -1) {
                        total += read;
                        if (total > maxBytes) throw new IllegalArgumentException("media download exceeds configured limit");
                        output.write(buffer, 0, read);
                    }
                    return output.toByteArray();
                }
            } catch (IOException exception) {
                throw new IllegalStateException("Unable to download media URL", exception);
            } catch (InterruptedException exception) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("Media download was interrupted", exception);
            }
        }
    }

    private static URI validate(URI value) {
        Objects.requireNonNull(value, "media URL must not be null");
        if (!value.isAbsolute() || !"https".equalsIgnoreCase(value.getScheme())
                || value.getHost() == null || value.getUserInfo() != null || value.getFragment() != null
                || value.toString().length() > 2048) {
            throw new IllegalArgumentException("media URL must be an HTTPS URL without credentials or fragments");
        }
        String host = value.getHost();
        if (host.equalsIgnoreCase("localhost") || host.endsWith(".localhost")) {
            throw new IllegalArgumentException("media URL host is not public");
        }
        try {
            for (InetAddress address : InetAddress.getAllByName(host)) {
                if (privateAddress(address)) throw new IllegalArgumentException("media URL target is private");
            }
        } catch (java.net.UnknownHostException exception) {
            throw new IllegalArgumentException("media URL host cannot be resolved", exception);
        }
        return value;
    }

    private static boolean privateAddress(InetAddress address) {
        if (address.isAnyLocalAddress() || address.isLoopbackAddress() || address.isLinkLocalAddress()
                || address.isSiteLocalAddress() || address.isMulticastAddress()) return true;
        if (address instanceof Inet6Address v6) {
            byte[] bytes = v6.getAddress();
            return (bytes[0] & 0xfe) == 0xfc; // IPv6 ULA (fc00::/7)
        }
        return false;
    }
}

File: qqbot-client/src/main/kotlin/com/mieai/qqbot/client/QqMediaDownloader.kt
package com.mieai.qqbot.client

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.Inet6Address
import java.net.InetAddress
import java.net.URI
import java.net.UnknownHostException
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage

/** Bounded, redirect-aware downloader used before handing media to QQ. */
class QqMediaDownloader(
    private val maxBytes: Long,
    private val timeout: Duration,
    private val maxRedirects: Int,
) {
    private val client: HttpClient

    init {
        require(maxBytes >= 1L) { "maxBytes must be positive" }
        require(!timeout.isZero && !timeout.isNegative) { "timeout must be positive" }
        require(maxRedirects in 0..8) { "maxRedirects is invalid" }
        client = HttpClient.newBuilder()
            .connectTimeout(timeout)
            .followRedirects(HttpClient.Redirect.NEVER)
            .build()
    }

    fun download(source: URI): CompletionStage<ByteArray> =
        CompletableFuture.supplyAsync { downloadSync(source) }

    private fun downloadSync(source: URI): ByteArray {
        var current = validate(source)
        var redirect = 0
        while (true) {
            val request = HttpRequest.newBuilder(current)
                .timeout(timeout)
                .header("Accept", "*/*")
                .GET()
                .build()
            try {
                val response = client.send(request, HttpResponse.BodyHandlers.ofInputStream())
                val status = response.statusCode()
                if (status in 300..399) {
                    response.body().close()
                    require(redirect < maxRedirects) { "media URL exceeded redirect limit" }
                    val location = response.headers().firstValue("location").orElseThrow {
                        IllegalArgumentException("media redirect has no Location header")
                    }
                    current = validate(current.resolve(location))
                    redirect++
                    continue
                }
                if (status !in 200..299) {
                    response.body().close()
                    throw IllegalArgumentException("media URL returned HTTP $status")
                }
                val declared = response.headers().firstValueAsLong("content-length").orElse(-1L)
                require(declared <= maxBytes) { "media download exceeds configured limit" }
                response.body().use { input ->
                    ByteArrayOutputStream().use { output ->
                        val buffer = ByteArray(8192)
                        var total = 0L
                        while (true) {
                            val read = input.read(buffer)
                            if (read == -1) break
                            total += read
                            require(total <= maxBytes) { "media download exceeds configured limit" }
                            output.write(buffer, 0, read)
                        }
                        return output.toByteArray()
                    }
                }
            } catch (exception: IOException) {
                throw IllegalStateException("Unable to download media URL", exception)
            } catch (exception: InterruptedException) {
                Thread.currentThread().interrupt()
                throw IllegalStateException("Media download was interrupted", exception)
            }
        }
    }

    companion object {
        private fun validate(value: URI): URI {
            require(value.isAbsolute && value.scheme.equals("https", ignoreCase = true) &&
                value.host != null && value.userInfo == null && value.fragment == null &&
                value.toString().length <= 2048) {
                "media URL must be an HTTPS URL without credentials or fragments"
            }
            val host = value.host
            require(!host.equals("localhost", ignoreCase = true) && !host.endsWith(".localhost")) {
                "media URL host is not public"
            }
            try {
                for (address in InetAddress.getAllByName(host)) {
                    require(!privateAddress(address)) { "media URL target is private" }
                }
            } catch (exception: UnknownHostException) {
                throw IllegalArgumentException("media URL host cannot be resolved", exception)
            }
            return value
        }

        private fun privateAddress(address: InetAddress): Boolean {
            if (address.isAnyLocalAddress || address.isLoopbackAddress || address.isLinkLocalAddress ||
                address.isSiteLocalAddress || address.isMulticastAddress) {
                return true
            }
            if (address is Inet6Address) {
                val bytes = address.address
                return (bytes[0].toInt() and 0xfe) == 0xfc
            }
            return false
        }
    }
}
